vector/src/main/java/im/vector/app/features/login/qr/QrCodeLoginViewModel.kt

```diff
@@ -0,0 +1,122 @@
+import kotlinx.coroutines.delay
+import kotlinx.coroutines.launch
+
+class QrCodeLoginViewModel @AssistedInject constructor(
+        @Assisted private val initialState: QrCodeLoginViewState,
+) : VectorViewModel<QrCodeLoginViewState, QrCodeLoginAction, QrCodeLoginViewEvents>(initialState) {
+
+    @AssistedFactory
+    interface Factory : MavericksAssistedViewModelFactory<QrCodeLoginViewModel, QrCodeLoginViewState> {
+        override fun create(initialState: QrCodeLoginViewState): QrCodeLoginViewModel
+    }
+
+    companion object : MavericksViewModelFactory<QrCodeLoginViewModel, QrCodeLoginViewState> by hiltMavericksViewModelFactory()
+
+    override fun handle(action: QrCodeLoginAction) {
+        when (action) {
+            is QrCodeLoginAction.OnQrCodeScanned -> handleOnQrCodeScanned(action)
+            QrCodeLoginAction.QrCodeViewStarted -> handleQrCodeViewStarted()
+            QrCodeLoginAction.ShowQrCode -> handleShowQrCode()
+        }
+    }
+
+    private fun handleShowQrCode() {
+        _viewEvents.post(QrCodeLoginViewEvents.NavigateToShowQrCodeScreen)
+    }
+
+    private fun handleQrCodeViewStarted() {
+        val qrCodeData = generateQrCodeData()
+        setState {
+            copy(
+                    generatedQrCodeData = qrCodeData
+            )
+        }
+    }
+
+    private fun handleOnQrCodeScanned(action: QrCodeLoginAction.OnQrCodeScanned) {
+        if (isValidQrCode(action.qrCode)) {
+            setState {
+                copy(
+                        connectionStatus = QrCodeLoginConnectionStatus.ConnectingToDevice
+                )
+            }
+            _viewEvents.post(QrCodeLoginViewEvents.NavigateToStatusScreen)
+        }
+
+        // TODO. UI test purpose. Fixme remove!
+        viewModelScope.launch {
+            delay(3000)
+            onFailed(QrCodeLoginErrorType.TIMEOUT, true)
+            delay(3000)
+            onConnectionEstablished("1234-ABCD-5678-EFGH")
+            delay(3000)
+            onSigningIn()
+            delay(3000)
+            onFailed(QrCodeLoginErrorType.DEVICE_IS_NOT_SUPPORTED, false)
+        }
+    }
+
+    private fun onFailed(errorType: QrCodeLoginErrorType, canTryAgain: Boolean) {
+        setState {
+            copy(
+                    connectionStatus = QrCodeLoginConnectionStatus.Failed(errorType, canTryAgain)
+            )
+        }
+    }
+
+    private fun onConnectionEstablished(securityCode: String) {
+        val canConfirmSecurityCode = initialState.loginType == QrCodeLoginType.LINK_A_DEVICE
+        setState {
+            copy(
+                    connectionStatus = QrCodeLoginConnectionStatus.Connected(securityCode, canConfirmSecurityCode)
+            )
+        }
+    }
+
+    private fun onSigningIn() {
+        setState {
+            copy(
+                    connectionStatus = QrCodeLoginConnectionStatus.SigningIn
+            )
+        }
+    }
+
+    /**
+     * TODO. UI test purpose. Fixme accordingly.
```

Comment:
Same question, will it be implemented later?

Reply:
Yes, but I want to keep placeholder functions.